### ISSUE TITLE
how to keep the login session instead of manually do this every time

### ISSUE THREAD

**Issue by nbg1rp** [3052/google#37]
I don't know why the session will be expired after a short time like several hours or so in my case.
After that, it will prompt as below:

msg=POST URL=https://android.googleapis.com/auth
panic: 403 Forbidden

if I re-do the login process, it will work again. Did I miss any steps or how can the session be last long time after login once?

**Comment by 3052** [comment]
sorry I dont know why it has the issue. if you know another tool that works as expected, I can make the change

**Comment by nbg1rp** [comment]
I did some search work for many similar repos, finally got this one:

Katya-Incorporated/Aurora-Store

perhaps you can check it for some hints, I have tried to install this apk and logined my account, it indeed download the apk  I specified. But I did not dig into the code of it. Hope this helps!

**Comment by 3052** [comment]
OK their login code is here:
https://github.com/Katya-Incorporated/Aurora-Store/blob/master/app/src/main/java/com/aurora/store/util/AC2DMTask.kt

but, the code I created already last 14 hours, how long yours last?

**Comment by nbg1rp** [comment]
I didn't get the exact time span, but it would be expired the next day I try. Is there any way to make this session token refreshed? Or how about this aurora work, could they make the session long live?

**Comment by 3052** [comment]
sorry but I need a timer, so I can know "pass" vs "fail" - so far mine work 15 hours

**Comment by nbg1rp** [comment]
please check AuthViewModel.kt line 60, buildGoogleAuthData, and the details may refer below location:

https://gitlab.com/AuroraOSS/gplayapi/-/blob/master/lib/src/main/java/com/aurora/gplayapi/helpers/AuthHelper.kt

**Comment by nbg1rp** [comment]
AuthViewModel.kt line 222, seems do another session generation when invalid.

**Comment by nbg1rp** [comment]
I have started the test  to detect the time consumption.

**Comment by 3052** [comment]
> AuthViewModel.kt line 222, seems do another session generation when invalid.

this line does not exist

**Comment by 3052** [comment]
> please check AuthViewModel.kt line 60, buildGoogleAuthData, and the details may refer below location:
> 
> https://gitlab.com/AuroraOSS/gplayapi/-/blob/master/lib/src/main/java/com/aurora/gplayapi/helpers/AuthHelper.kt

this is not correct, because that function takes `aasToken`. the concern is generating `aasToken` that last longer

> AuthViewModel.kt line 222, seems do another session generation when invalid.

this file does not exist. dont waste my fucking time. I am closing this for now. get your shit together and I will reopen.

**Comment by 3052** [comment]
> please check AuthViewModel.kt line 60, buildGoogleAuthData, and the details may refer below location:
> 
> https://gitlab.com/AuroraOSS/gplayapi/-/blob/master/lib/src/main/java/com/aurora/gplayapi/helpers/AuthHelper.kt

this is not correct, because that function takes `aasToken`. the concern is generating `aasToken` that last longer

> AuthViewModel.kt line 222, seems do another session generation when invalid.

dont do that again. if you have a file link to it. you've been warned. 

https://github.com/Katya-Incorporated/Aurora-Store/blob/master/app/src/main/java/com/aurora/store/viewmodel/auth/AuthViewModel.kt

**Comment by nbg1rp** [comment]
so what it does just like your implementation, right? I saw it use webview to show the login UI, perhaps we can use selenium to get the auth cookie we need instead of manually operating.

**Comment by 3052** [comment]
my token working for 18 hours so far

> so what it does just like your implementation, right?

this has some differences:
https://github.com/Katya-Incorporated/Aurora-Store/blob/master/app/src/main/java/com/aurora/store/util/AC2DMTask.kt

so I could modify my code, but again we need to know how long my token last.

> I saw it use webview to show the login UI, perhaps we can use selenium to get the auth cookie we need instead of manually operating.

I want to use webView, but all the implementations suck currently:

1. https://github.com/jchv/go-webview2/issues/72
2. https://github.com/gioui-plugins/gio-plugins/issues/59
3. https://github.com/inkeliz/gowebview/issues/25
4. https://github.com/inkeliz/giowebview/issues/10
5. https://github.com/suchipi/webview/issues/6
6. https://github.com/CarsonSlovoka/go-webview2/issues/5
7. https://github.com/wailsapp/go-webview2/issues/5
8. https://github.com/mekkanized/go-webview/issues/3
9. https://github.com/isNeilLin/LightApp/issues/1

**Comment by 3052** [comment]
requires CGO :(

https://github.com/webview/webview_go/blob/master/webview.go#L4-L24

**Comment by 3052** [comment]
> My session is expired now, lasting about 15~16 hours. I have successfully query app details one hour ago.

mine is still working, 38 hours

> And the auro store seems update its session(creating new work environment or so...), it can still use the account without do manually login.

OK, how?

> please double check this, I think it's the key part, in its gplayapi, it will re-auth.

where is the re-auth code?

> requires CGO :(

I might be able to have GUI login as optional, if people want that. or they can still do cookie manually. you can use `zig cc` for CGO:

https://ziglang.org/download

**Comment by 3052** [comment]
> My session is expired now, lasting about 15~16 hours. I have successfully query app details one hour ago.

mine is still working, 38 hours

> And the auro store seems update its session(creating new work environment or so...), it can still use the account without do manually login.

OK, how?

> please double check this, I think it's the key part, in its gplayapi, it will re-auth.

where is the re-auth code?

> perhaps you can try this: https://github.com/webview/webview_go/blob/master/webview.go
> 
> Init() a js first, and Bind() a go function to receive cookie, injected js set up a callback when page loaded to send cookie to the binded function.

the cookie is HttpOnly, so you cannot get it with JavaScript:

<https://developer.mozilla.org/docs/Web/HTTP/Cookies#restrict_access_to_cookies>

**Comment by 3052** [comment]
why you close?

**Comment by liconway** [comment]
I just saw this issue through the mention. `mekkanized/go-webview` is more of a PoC than a production ready tool as of now, where the goal is to have webview functionality without using CGO. I'm hoping to build it up in the future, though I just don't have the time right now.

That being said, the same process mentioned here for `webview_go` should be applicable to `go-webview`:

> perhaps you can try this: https://github.com/webview/webview_go/blob/master/webview.go
> 
> Init() a js first, and Bind() a go function to receive cookie, injected js set up a callback when page loaded to send cookie to the binded function.

No guarantees about functionality or consistency until I can develop it further though, so UAYOR

**Comment by 3052** [comment]
> That being said, the same process mentioned here for `webview_go` should be applicable to `go-webview`:

as I already said, the cookie in question is HttpOnly, so you cannot get it with JavaScript:

https://developer.mozilla.org/docs/Web/HTTP/Cookies#restrict_access_to_cookies

**Comment by 3052** [comment]
> That being said, the same process mentioned here for `webview_go` should be applicable to `go-webview`:

@liconway  as I already said, the cookie in question is HttpOnly, so you cannot get it with JavaScript:

https://developer.mozilla.org/docs/Web/HTTP/Cookies#restrict_access_to_cookies

**Comment by liconway** [comment]
@3052 Ah apologies, I missed that. There's no good way I can think of doing at the moment then, sorry about that. I'll update the corresponding issue in the future if/when that changes

**Comment by 3052** [comment]
@liconway there is a method here, but the API is really awkward:

https://github.com/zzl/go-webview2/issues/1
